Sources/TuistGenerator/Generator/KnownAssetTagsGenerator.swift

```diff
@@ -0,0 +1,70 @@
+import Foundation
+import TuistGraph
+
+private let assetFormats: Set = [
+    "arresourcegroup",
+    "cubetextureset",
+    "dataset",
+    "imageset",
+    "imagestack",
+    "spriteatlas",
+    "textureset",
+]
+
+private struct ContentJson: Decodable {
+    struct ContentProperties: Decodable {
+        enum CodingKeys: String, CodingKey {
+            case onDemandResourceTags = "on-demand-resource-tags"
+        }
+
+        let onDemandResourceTags: [String]
+    }
+
+    let properties: ContentProperties
+}
+
+protocol KnownAssetTagsGenerating: AnyObject {
+    func generate(project: Project) throws -> [String]
+}
+
+final class KnownAssetTagsGenerator: KnownAssetTagsGenerating {
+    func generate(project: Project) throws -> [String] {
+        var tags = project.targets.map { $0.resources.resources.map(\.tags).flatMap { $0 } }.flatMap { $0 }
+
+        let initialInstallTags = project.targets.compactMap {
+            $0.onDemandResourcesTags?.initialInstall?.compactMap { $0 }
+        }.flatMap { $0 }
+
+        let prefetchOrderTags = project.targets.compactMap {
+            $0.onDemandResourcesTags?.prefetchOrder?.compactMap { $0 }
+        }.flatMap { $0 }
+
+        tags.append(contentsOf: initialInstallTags)
+        tags.append(contentsOf: prefetchOrderTags)
+
+        var assetsTags: [String] = []
+        let decoder = JSONDecoder()
+        for target in project.targets {
+            for resource in target.resources.resources {
+                guard let children = try? resource.path.path.recursiveChildren() else { continue }
+                for child in children {
+                    guard let pathExtension = child.extension, assetFormats.contains(pathExtension) else { continue }
+                    guard let contents = try? child.children().first(where: {
+                        $0.lastComponent == "Contents.json"
+                    }) else {
+                        continue
+                    }
+                    guard let data = try? Data(contentsOf: contents.url) else { continue }
```

Comment:
@tuist/core do we have other examples where we perform loading during generation - from memory we always made an effort to offload all of these to the loading step - can't think of better alternatives for the moment so may be ok.